modules/core/src/test/java/org/apache/ignite/internal/processors/cache/transactions/TxRecoveryConcurrentTest.java

```diff
@@ -0,0 +1,169 @@
+
+        stopAllGrids();
+    }
+
+    /**
+     * The test enforces the concurrent processing of the same prepared transaction
+     * both in the tx recovery procedure started due to near and primary node left and in the
+     * tx recovery request handler invoked by message from another backup node.
+     * <p>
+     * The idea is to have a 3-nodes cluster and a cache with 2 backups. So there
+     * will be 2 backup nodes to execute the tx recovery in parallel if primary one
+     * would fail. These backup nodes will send the tx recovery requests to each
+     * other, so the tx recovery request handler will be invoked as well.
+     * <p>
+     * Use several attempts to reproduce the race condition.
+     * <p>
+     * Expected result: transaction is finished on both backup nodes and the partition
+     * map exchange is completed as well.
+     */
+    @Test
+    public void testRecoveryNotDeadLockOnNearAndPrimaryFail() throws Exception {
+        final IgniteEx backup = startGrids(2);
+
+        final CyclicBarrier backupBlockerBarrier = new CyclicBarrier(3);
+
+        final Runnable backupBlockerTask = () -> {
+            try {
+                backupBlockerBarrier.await();
+                backupBlockerBarrier.await();
+            }
+            catch (InterruptedException | BrokenBarrierException ignored) {
+                // Just supress.
+            }
+        };
+
+        for (int iter = 0; iter < 100; iter++) {
+            final IgniteEx primary = startGrid(2);
+
+            awaitPartitionMapExchange();
+
+            final IgniteCache<Object, Object> cache = primary.cache(DEFAULT_CACHE_NAME);
+
+            final TransactionProxyImpl<?, ?> tx = (TransactionProxyImpl<?, ?>)primary.transactions().txStart();
+
+            final Integer key = primaryKeys(cache, 1, 0).get(0);
+
+            cache.put(key, key);
+
+            tx.tx().prepare(true);
+
+            for (Ignite grid : G.allGrids())
+                assertTrue(((IgniteEx)grid).context().cache().context().tm().activeTransactions().size() == 1);
+
+            Collection<IgniteInternalTx> backupTransactions = new LinkedList<>();
+
+            for (Ignite grid : backupNodes(key, DEFAULT_CACHE_NAME))
+                backupTransactions.addAll(((IgniteEx)grid).context().cache().context().tm().activeTransactions());
+
+            assertTrue(backupTransactions.size() == 2);
+
+            // Block recovery procedure processing on one of the backup nodes.
+            backup.context().pools().getSystemExecutorService().execute(backupBlockerTask);
+
+            // Block stripe tx recovery request processing on one of the backup nodes (note that the only stripe is
+            // configured in the executor).
+            backup.context().pools().getStripedExecutorService().execute(0, backupBlockerTask);
+
+            // Prevent tx finish request processing on both backup nodes.
+            for (Ignite grid : backupNodes(key, DEFAULT_CACHE_NAME))
```

Comment:
Any reason to recalculate `backupNodes()`?

Reply:
Reworked